openshift-kube-apiserver/authorization/minimumkubeletversion/minimum_kubelet_version.go

```diff
@@ -0,0 +1,93 @@
+package minimumkubeletversion
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	openshiftfeatures "github.com/openshift/api/features"
+	nodelib "github.com/openshift/library-go/pkg/apiserver/node"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/util/feature"
+	v1listers "k8s.io/client-go/listers/core/v1"
+	cache "k8s.io/client-go/tools/cache"
+	"k8s.io/component-base/featuregate"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/auth/nodeidentifier"
+)
+
+type minimumKubeletVersionAuth struct {
+	nodeIdentifier          nodeidentifier.NodeIdentifier
+	nodeInformer            cache.SharedIndexInformer
+	nodeLister              v1listers.NodeLister
+	minVersion              *semver.Version
+	hasNodeInformerSyncedFn func() bool // factored for unit tests
+}
+
+// Creates a new minimumKubeletVersionAuth object, which is an authorizer that checks
+// whether nodes are new enough to be authorized.
+func NewMinimumKubeletVersion(minVersion *semver.Version,
+	nodeIdentifier nodeidentifier.NodeIdentifier,
+	nodeInformer cache.SharedIndexInformer,
+	nodeLister v1listers.NodeLister,
+) *minimumKubeletVersionAuth {
+	if !feature.DefaultMutableFeatureGate.Enabled(featuregate.Feature(openshiftfeatures.FeatureGateMinimumKubeletVersion)) {
+		minVersion = nil
+	}
+
+	return &minimumKubeletVersionAuth{
+		nodeIdentifier:          nodeIdentifier,
+		nodeInformer:            nodeInformer,
+		nodeLister:              nodeLister,
+		hasNodeInformerSyncedFn: nodeInformer.HasSynced,
+		minVersion:              minVersion,
+	}
+}
+
+func (m *minimumKubeletVersionAuth) Authorize(ctx context.Context, attrs authorizer.Attributes) (authorizer.Decision, string, error) {
+	if m.minVersion == nil {
+		return authorizer.DecisionNoOpinion, "", nil
+	}
+
+	nodeName, isNode := m.nodeIdentifier.NodeIdentity(attrs.GetUser())
+	if !isNode {
+		// ignore requests from non-nodes
+		return authorizer.DecisionNoOpinion, "", nil
+	}
+
+	if len(nodeName) == 0 {
+		return authorizer.DecisionNoOpinion, fmt.Sprintf("unknown node for user %q", attrs.GetUser().GetName()), nil
+	}
+
+	// Short-circut if "subjectaccessreviews", or a "get" or "update" on the node object.
+	// Regardless of kubelet version, it should be allowed to do these things.
+	if attrs.IsResourceRequest() {
+		requestResource := schema.GroupResource{Group: attrs.GetAPIGroup(), Resource: attrs.GetResource()}
+		switch requestResource {
+		case api.Resource("nodes"):
+			if v := attrs.GetVerb(); v == "get" || v == "update" {
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+		// TODO(haircommander): do we need other flavors of access reviews here?
```

Comment:
Good question.

I think that `SubjectAccessReview` allows for checking access for any user, `LocalSubjectAccessReview` for checking access for any user in a namespace and `SelfSubjectAccessReview` for checking access for the the current user. 

Do we know which API `kubelet` use/might use ?

Reply:
I am struggling to find the code in the kubelet where this actually happens. If I were to write it I'd say SelfSubjectAccessReview is most fitting. I did find a note in the kubelet webhook documentation that SubjectAccessReview API is used. @deads2k : allowing SAR was recommended by you, do you happen to know?

Reply:
[added](https://github.com/haircommander/kubernetes/compare/68184726bad8affbf2bbbfa70cd3b73434a82c32..1e7863e04c1b24ab6a7d6e36a5dcae765db93542) self just in case, local didn't seem right